File: matlab/registration_example.m
CPWD = pwd;
SANDBOX_NAME = 'gira3d-registration';
matches = split(CPWD, SANDBOX_NAME);
GIRA3D_REGISTRATION_SANDBOX = [matches{1}, SANDBOX_NAME];
DATA_DIR = [GIRA3D_REGISTRATION_SANDBOX, '/data/rgbd_dataset_freiburg3_long_office_household/100_components/'];
PCLD_DIR = [GIRA3D_REGISTRATION_SANDBOX, '/data/rgbd_dataset_freiburg3_long_office_household/pointclouds/'];
RESULTS_DIR = [DATA_DIR, 'results/'];

target_file = [DATA_DIR, '1200.gmm'];
source_file = [DATA_DIR, '1260.gmm'];

target_pcld = load([PCLD_DIR, '1200.txt']);
source_pcld = load([PCLD_DIR, '1260.txt']);

if ~exist(RESULTS_DIR, 'dir')
  mkdir(RESULTS_DIR)
end

%%%%%%%%%%% Isoplanar Registration
x_opt = zeros(6,1);
[x_opt, score] = isoplanar_registration(source_file, target_file, zeros(6,1));
assert(max(abs(x_opt - [-0.7795; -0.2672; 0.5195; -0.0513; 0.5727; 0.3276])) < 0.001)

%%%%%%%%%%% Anisotropic Registration
[x_opt, score] = anisotropic_registration(source_file, target_file, x_opt);
assert(max(abs(x_opt - [-0.4217; -0.2275; 0.4124; -0.0441; 0.7002; 0.4001])) < 0.001)
Rotation = axang2rotm([x_opt(4:6) / norm(x_opt(4:6)); norm(x_opt(4:6))]');
translation = x_opt(1:3);
disp('Tests passed')

%%%%%%%%%%% Visualization
gmm_target = GMM3();
gmm_target.load(target_file);

gmm_source = GMM3();
gmm_source.load(source_file);

figure;
hold on; pcshow(target_pcld);
colormap(flipud(jet));
hold on; pcshow(source_pcld);
colormap(flipud(jet));
title('Before Registration');
axis equal;
view(0, -90);

figure;
hold on; pcshow(target_pcld);
colormap(flipud(jet));
source_pcld_transformed = Rotation * transpose(source_pcld) + repmat(translation, 1, size(source_pcld,1));
hold on; pcshow(source_pcld_transformed');
colormap(flipud(jet));
title('After Registration');
axis equal;
view(0, -90);

File: python/registration_example.py
#!/usr/bin/env python
import copy
import numpy as np
import os
from os.path import expanduser
import open3d
#from sogmm_py.utils import np_to_open3d

import gmm_d2d_registration_py
from utils.RToZYX import RToZYX
from utils.open3d_visualizer import Open3DVisualizer

def load_np_from_txt(txt_file):
    data = np.loadtxt(txt_file, delimiter=',')
    return data

def main():

    cwd = os.getcwd()
    SANDBOX_NAME = 'gira3d-registration'
    matches = cwd.split(SANDBOX_NAME)
    GIRA3D_REGISTRATION_SANDBOX = matches[0] + SANDBOX_NAME
    DATA_DIR = GIRA3D_REGISTRATION_SANDBOX + '/data/rgbd_dataset_freiburg3_long_office_household/100_components/'
    PCLD_DIR = GIRA3D_REGISTRATION_SANDBOX + '/data/rgbd_dataset_freiburg3_long_office_household/pointclouds/'
    RESULTS_DIR = DATA_DIR + 'results/'

    source_idx = 1260
    target_idx = 1200

    source_file = DATA_DIR + str(source_idx) + '.gmm'
    target_file = DATA_DIR + str(target_idx) + '.gmm'

    Tinit = np.eye(4)
    output = gmm_d2d_registration_py.isoplanar_registration(Tinit, source_file, target_file)
    ret = gmm_d2d_registration_py.anisotropic_registration(output[0], source_file, target_file)
    Tout = ret[0]
    translation = (Tout[0:3,3]).flatten()
    correct_translation = np.array([-0.4217, -0.2275, 0.4124])
    assert(np.max(np.abs(translation-correct_translation)) < 1e-3)

    correct_zyx = np.array([0.1211, 0.6879, 0.4607])
    assert(np.max(np.abs(correct_zyx - RToZYX(Tout[0:3,0:3]))) < 1e-3)
    print('Tests Passed!')

    source_pcld_file = PCLD_DIR + str(source_idx) + '.txt'
    target_pcld_file = PCLD_DIR + str(target_idx) + '.txt'

    viz = Open3DVisualizer()
    source_np = load_np_from_txt(source_pcld_file)
    target_np = load_np_from_txt(target_pcld_file)
    np_array = np.concatenate([source_np, target_np])
    viz.plot3d(np_array)

    Rotation = Tout[0:3, 0:3]
    t = np.reshape(translation, (3,1))
    ts = np.tile(t, (1, np.shape(source_np)[0]))
    rotated_source = np.matmul(Rotation, source_np.transpose())
    transformed_source = rotated_source + ts
    np_array = np.concatenate([transformed_source.transpose(), target_np])
    viz.plot3d(np_array)

if __name__ == '__main__':
  main()
